dali/kernels/signal/fft/fft_cpu_test.cc

```diff
@@ -0,0 +1,444 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include <tuple>
+#include <vector>
+#include <complex>
+#include <cmath>
+#include "dali/kernels/scratch.h"
+#include "dali/kernels/signal/fft/fft_cpu.h"
+#include "dali/test/test_tensors.h"
+#include "dali/test/tensor_test_utils.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+namespace fft {
+namespace test {
+
+void NaiveDft(std::vector<std::complex<float>> &out,
+              const span<const float>& in,
+              int64_t nfft,
+              bool full_spectrum = false) {
+  auto n = in.size();
+  auto out_size = full_spectrum ? nfft : nfft/2+1;
+  out.clear();
+  out.reserve(out_size);
+  // Loop through each sample in the frequency domain
+  for (int64_t k = 0; k < out_size; k++) {
+    float real = 0.0f, imag = 0.0f;
+    // Loop through each sample in the time domain
+    for (int64_t i = 0; i < n; i++) {
+      auto x = in[i];
+      real += x * cos(2.0f * M_PI * k * i / n);
+      imag += -x * sin(2.0f * M_PI * k * i / n);
+    }
+    out.push_back({real, imag});
+  }
+}
+
+void CompareFfts(const span<const float>& reference,
+                 const span<const float>& fft,
+                 float eps = 1e-3) {
+  ASSERT_EQ(reference.size(), fft.size());
+  const auto *ref = reference.data(), *data = fft.data();
+  for (int i = 0; i < fft.size(); i++) {
+    auto diff = ref[i] - data[i];
+    auto diff_max = ref[i] * eps;
+    // Error tends to be big if the numbers are big
+    if (diff_max < eps)
+      diff_max = eps;
+    ASSERT_LE(diff, diff_max);
+  }
+}
+
+void CompareFfts(const span<const std::complex<float>>& reference,
+                 const span<const std::complex<float>>& fft,
+                 float eps = 1e-3) {
+  span<const float> ref_f = {reinterpret_cast<const float*>(reference.data()), reference.size()*2};
+  span<const float> fft_f = {reinterpret_cast<const float*>(fft.data()), fft.size()*2};
+  CompareFfts(ref_f, fft_f, eps);
+}
+
+void PowerSpectrum(span<float> out,
+                   span<std::complex<float>> in,
+                   int64_t nfft,
+                   float floor_pow = 1e-30) {
```

Comment:
Not there any more in the implementation?

Reply:
done